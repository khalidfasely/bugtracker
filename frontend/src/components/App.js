function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
