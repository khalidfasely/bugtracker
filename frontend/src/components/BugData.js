const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;