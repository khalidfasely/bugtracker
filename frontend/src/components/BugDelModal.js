import Modal from "react-modal";

const BugDelModal = ({ delModalOpen, setDelModalOpen, deleteBug }) => (
    <Modal
        isOpen={delModalOpen}
        onRequestClose={() => setDelModalOpen(false)}
    >
        If you delete this bug there is no way to return it!
        <button onClick={deleteBug} data-testid='delete_button'>Delete</button>
    </Modal>
);

export default BugDelModal;