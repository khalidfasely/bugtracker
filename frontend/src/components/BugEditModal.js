import Modal from "react-modal";

import NewBug from "./NewBug";

const BugEditModal = ({
    editModalOpen, setEditModalOpen,
    users, bug
}) => (
    <Modal
        isOpen={editModalOpen}
        onRequestClose={() => setEditModalOpen(false)}
    >
        <NewBug users={users} isEdit={true} bug={bug} setEditModalOpen={setEditModalOpen} />
    </Modal>
);

export default BugEditModal;