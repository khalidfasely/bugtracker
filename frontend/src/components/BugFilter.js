import { connect } from "react-redux";
import { setTextFilter, sortByDefault, sortByPriority } from "../actions/bugFilter";

export const BugFilter = ({ filters, setTextFilter, sortByDefault, sortByPriority }) => {
    const onSortByChange = (e) => {
        if (e.target.value === 'priority') {
            sortByPriority();
            return;
        };
        sortByDefault();
    };

    return (
        <div className="bugs-filter">
            <div className="input-group__item">
                <input
                    data-testid='search_input'
                    type="text"
                    className="text-input"
                    placeholder="Search bugs"
                    value={filters.text}
                    maxLength={35}
                    onChange={(e) => setTextFilter(e.target.value)}
                />
            </div>
            <div className="input-group__item">
                <select
                    data-testid='select_input'
                    className="select"
                    value={filters.sortBy}
                    onChange={onSortByChange}
                >
                    <option value="">Date</option>
                    <option value="priority">Priority</option>
                </select>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => ({
    filters: state.bugFilter
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByPriority: () => dispatch(sortByPriority()),
    sortByDefault: () => dispatch(sortByDefault())
});

export default connect(mapStateToProps, mapDispatchToProps)(BugFilter);