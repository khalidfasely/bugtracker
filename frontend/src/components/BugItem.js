import { Link } from "react-router-dom";
import { connect } from "react-redux";

import Comment from "./Comment";

const BugItem = ({uname, bug, comments}) => {
    console.log(bug, comments);
    return (
        <>
            {
                bug?.users_with?.includes(uname) ?
                <div>
                    **** bug <br /><h1>{bug.title}</h1>
                    **** comments {comments?.map(comment => <Comment comment={comment} />)}
                </div> :
                <div>Not able to see this bug! <Link to="/new">Create one!</Link></div>
            }
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    bug: state.bugItem.bug,
    comments: state.bugItem.comments
});

export default connect(mapStateToProps)(BugItem);