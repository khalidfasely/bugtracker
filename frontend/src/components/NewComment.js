import { useState } from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";

import { startSetNewComment } from "../actions/bug";

const NewComment = ({ bugId, uname, startSetNewComment, commentEdit, setIsEdit }) => {
    const [comment, setComment] = useState(commentEdit ? commentEdit : '');
    const [error, setError] = useState('');

    const handleSubmit = (e) => {
        e.preventDefault();

        setError('');

        if (comment.replace(/\s/g, '') < 25) {
            setError('Comment must be 25 character or more!');
            return;
        };

        if (commentEdit) {
            console.log('edited', comment);
            //startSetEditComment(comment);
            setIsEdit(false);
            return;
        }

        startSetNewComment(bugId, comment);

        setComment('');
    };

    if (!uname) {
        return <div><Link to='/login'>Login</Link> or <Link to='/register'>Sign In</Link> to Add a Comment!</div>;
    }

    return (
        <div>
            <form onSubmit={handleSubmit}>
                {error && <p>{error}</p>}
                <textarea
                    placeholder="Create A Comment on this Bug!"
                    maxLength={255}
                    value={comment} onChange={(e) => setComment(e.target.value)}
                />
                {
                    commentEdit ?
                    <button>Edit</button> :
                    <button>Add</button>
                }
            </form>
        </div>
    );
};

const mapStateToProps = (state) => ({
    uname: state.auth.uname
});

const mapDispatchToProps = (dispatch) => ({
    startSetNewComment: (bugId, comment) => dispatch(startSetNewComment(bugId, comment))
});

export default connect(mapStateToProps, mapDispatchToProps)(NewComment);