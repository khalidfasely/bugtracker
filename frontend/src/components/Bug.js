import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);