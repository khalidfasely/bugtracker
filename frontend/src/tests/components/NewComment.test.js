import React from "react";
import { render, fireEvent } from '@testing-library/react';

import { NewComment } from "../../components/NewComment";

let newCommentProps, editCommentProps;

beforeEach(() => {
    newCommentProps = {
        bugId: 1,
        uname: 'admin',
        startSetNewComment: jest.fn(() => Promise.resolve())
    }
    editCommentProps = {
        commentEdit: 'You can fix this bug by adding a default value!',
        commentId: 1,
        setIsEdit: jest.fn(),
        startSetEditComment: jest.fn(() => Promise.resolve())
    }
});

test('Should render NewComment component correctly without props', () => {
    const { asFragment } = render(<NewComment />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render NewComment component correctly with props to new comment', () => {
    const { asFragment } = render(<NewComment {...newCommentProps} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render NewComment component correctly with props to edit comment', () => {
    const { asFragment } = render(<NewComment {...{...newCommentProps, ...editCommentProps}} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should change comment input correctly in new comment', () => {
    const { getByTestId } = render(<NewComment {...newCommentProps} />);
    const commentInputEl = getByTestId('comment_input');

    expect(commentInputEl.value).toBe('');

    fireEvent.change(commentInputEl, {
        target: { value: 'New comment!' }
    });

    expect(commentInputEl.value).toBe('New comment!')
});

test('Should change comment input correctly in edit comment', () => {
    const { getByTestId } = render(<NewComment {...{...newCommentProps, ...editCommentProps}} />);
    const commentInputEl = getByTestId('comment_input');

    expect(commentInputEl.value).toBe(editCommentProps.commentEdit);

    fireEvent.change(commentInputEl, {
        target: { value: 'New comment!' }
    });

    expect(commentInputEl.value).toBe('New comment!')
});

test('Should display an error if comment length < 15', () => {
    const { getByTestId } = render(<NewComment {...newCommentProps} />);
    const formEl = getByTestId('comment_form');
    const commentInputEl = getByTestId('comment_input');
    const errorEl = getByTestId('comment_error');

    expect(errorEl.textContent).toBe('');

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    expect(errorEl.textContent).toBe('Comment must be 15 character or more!');

    fireEvent.change(commentInputEl, {
        target: { value: 'New comment for erro to gone!' }
    });

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    expect(errorEl.textContent).toBe('');
});

test('Should fire startSetNewComment correctly on form submit', () => {
    const { getByTestId } = render(<NewComment {...newCommentProps} />);
    const formEl = getByTestId('comment_form');
    const commentInputEl = getByTestId('comment_input');

    fireEvent.change(commentInputEl, {
        target: { value: 'You can fix this bug by adding a default value!' }
    });

    expect(newCommentProps.startSetNewComment).toHaveBeenCalledTimes(0);

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    expect(newCommentProps.startSetNewComment).toHaveBeenCalledTimes(1);
});

test('Should fire startSetEditComment correctly on form submit', () => {
    const { getByTestId } = render(<NewComment {...{...newCommentProps, ...editCommentProps}} />);
    const formEl = getByTestId('comment_form');

    expect(editCommentProps.startSetEditComment).toHaveBeenCalledTimes(0);

    fireEvent.submit(formEl, {
        preventDefault: () => {}
    });

    expect(editCommentProps.startSetEditComment).toHaveBeenCalledTimes(1);
});