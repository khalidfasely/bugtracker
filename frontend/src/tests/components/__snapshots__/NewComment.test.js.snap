// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render NewComment component correctly with props to edit comment 1`] = `
<DocumentFragment>
  <div
    class="comment-form-container"
  >
    <form
      class="comment-form"
      data-testid="comment_form"
    >
      <p
        class="error_message"
        data-testid="comment_error"
      />
      <textarea
        data-testid="comment_input"
        maxlength="255"
        placeholder="Create A Comment on this Bug!"
      >
        You can fix this bug by adding a default value!
      </textarea>
      <div
        class="buttons-container"
      >
        <button>
          Save
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Should render NewComment component correctly with props to new comment 1`] = `
<DocumentFragment>
  <div
    class="comment-form-container"
  >
    <form
      class="comment-form"
      data-testid="comment_form"
    >
      <p
        class="error_message"
        data-testid="comment_error"
      />
      <textarea
        data-testid="comment_input"
        maxlength="255"
        placeholder="Create A Comment on this Bug!"
      />
      <div
        class="buttons-container"
      >
        <button
          disabled=""
        >
          Add
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Should render NewComment component correctly without props 1`] = `
<DocumentFragment>
  <div>
    <a>
      Login
    </a>
     or 
    <a>
      Sign In
    </a>
     to Add a Comment!
  </div>
</DocumentFragment>
`;
